core/near-network/src/lib.rs

```diff
@@ -0,0 +1,439 @@
+        peer_id: PeerId,
+        account_id: Option<AccountId>
+    ) {
+        let handshake_msg = ServiceEvent::HandShake { 
+            peer_id,
+            account_id,
+        };
+        tokio::spawn(
+            sender.send(handshake_msg).map(|_| ()).map_err(|e| {
+                error!("Error when sending handshake message: {:?}", e);
+            })
+        );
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub enum ServiceEvent {
+    /// remote closed connection
+    NodeClosed {
+        peer_id: PeerId,
+    },
+    /// handshake message
+    HandShake {
+        peer_id: PeerId,
+        account_id: Option<AccountId>,
+    },
+    /// custom message
+    Message {
+        peer_id: PeerId,
+        data: Vec<u8>
+    },
+    /// account info received from peer
+    AccountInfo {
+        peer_id: PeerId,
+        info: HashMap<AccountId, PeerId>,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ::primitives::hash::hash_struct;
+    use std::thread;
+    use std::time::Duration;
+    use std::sync::Arc;
+    use parking_lot::Mutex;
+
+    #[test]
+    fn test_two_peers() {
+        let addr1 = "127.0.0.1:3000";
+        let addr2 = "127.0.0.1:3001";
+        let peer_id1 = hash_struct(&0);
+        let peer_id2 = hash_struct(&1);
+        let (message_tx1, _) = channel(1024);
+        let (message_tx2, _) = channel(1024);
+        let service1 = Arc::new(Mutex::new(Service::new(addr1.clone(), peer_id1, message_tx1)));
+        let service2 = Arc::new(Mutex::new(Service::new(addr2.clone(), peer_id2, message_tx2)));
+        let peer = Peer::new(addr1.parse::<SocketAddr>().unwrap(), peer_id1, None);
+        
+        let task = futures::lazy({
+            let service1 = service1.clone();
+            let service2 = service2.clone();
+            move || {
+                service1.lock().spawn_listening_task();
+                service2.lock().spawn_listening_task();
+                service2.lock().dial(peer.addr);
+                Ok(())
+            }
+        });
+        thread::spawn(move || tokio::run(task));
+        while service1.lock().connected_peers.read().len() < 1 
+            || service2.lock().connected_peers.read().len() < 1 {
+            thread::sleep(Duration::from_secs(1));
+        }
+    }
+
+    #[test]
+    fn test_send_message() {
+        let addr1 = "127.0.0.1:3002";
+        let addr2 = "127.0.0.1:3003";
+        let peer_id1 = hash_struct(&0);
+        let peer_id2 = hash_struct(&1);
+        let (message_tx1, _) = channel(1024);
+        let (message_tx2, message_rx2) = channel(1024);
+        let service1 = Arc::new(Mutex::new(Service::new(addr1.clone(), peer_id1, message_tx1)));
+        let service2 = Arc::new(Mutex::new(Service::new(addr2.clone(), peer_id2, message_tx2)));
+        let peer = Peer::new(addr1.parse::<SocketAddr>().unwrap(), peer_id1, None);
+        let message_queue = Arc::new(Mutex::new(vec![]));
+        thread::spawn({
```

Comment:
I don't want the test to depend on time. It's better to wait on some specific condition.

Reply:
There are two points:
* We can try using `wait()` instead of adding an additional queue, because it is shorter.
* We should have a timeout on the test. Right now the test would hang if message_queue is not populated due to some bug.